src/main/java/com/uber/cadence/internal/worker/Poller.java

```diff
@@ -0,0 +1,206 @@
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+package com.uber.cadence.internal.worker;
+
+import com.uber.cadence.internal.common.BackoffThrottler;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Calls a passed task in a loop according to {@link PollerOptions}.
+ */
+final class Poller implements SuspendableWorker {
+
+    interface ThrowingRunnable {
+        void run() throws Throwable;
+    }
+
+    private class PollServiceTask implements Runnable {
+
+        private final ThrowingRunnable task;
+
+        PollServiceTask(ThrowingRunnable task) {
+            this.task = task;
+        }
+
+        @Override
+        public void run() {
+            try {
+                if (log.isDebugEnabled()) {
+                    log.debug("poll task begin");
+                }
+
+                if (pollExecutor.isTerminating()) {
+                    return;
+                }
+                pollBackoffThrottler.throttle();
+                if (pollExecutor.isTerminating()) {
+                    return;
+                }
+                if (pollRateThrottler != null) {
+                    pollRateThrottler.throttle();
+                }
+
+                CountDownLatch suspender = Poller.this.suspendLatch.get();
+                if (suspender != null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("poll task suspending latchCount=" + suspender.getCount());
+                    }
+                    suspender.await();
+                }
+
+                if (pollExecutor.isTerminating()) {
+                    return;
+                }
+                task.run();
+                pollBackoffThrottler.success();
+            } catch (Throwable e) {
+                pollBackoffThrottler.failure();
+                if (!(e.getCause() instanceof InterruptedException)) {
+                    uncaughtExceptionHandler.uncaughtException(Thread.currentThread(), e);
+                }
+            } finally {
+                // Resubmit itself back to pollExecutor
+                if (!pollExecutor.isShutdown()) {
+                    pollExecutor.execute(this);
```

Comment:
this is weird.. I am sure there is a reason for doing this (rather than a simple while loop)